torch/onnx/symbolic_opset9.py

```diff
@@ -1477,7 +1438,10 @@ def glu(g: jit_utils.GraphContext, input, dim):
     if dim_size is not None:
         assert dim_size % 2 == 0
 
-    first, second = g.op("Split", input, axis_i=dim, outputs=2)
+    if g.opset < 18:
```

Comment:
This comment applies to other ops that are doing similar opset checks

Ideally, `torch/onnx/symbolic_opset9.py::glu` should not do `if g.opset < 18` checks. Instead, `symbolic_opset9.py` should only have code for glu opset 9 and `symbolic_opset20.py` should have code for glu opset 20. tracking this is very hard because we will look into one file and the actual implementation will be elsewhere hidden by a conditional

Maybe the code could look as follow:

torch/onnx/symbolic_opset9.py
```python
 @_onnx_symbolic("aten::glu")
@symbolic_helper.parse_args("v", "i")
@_beartype.beartype
def glu(g: jit_utils.GraphContext, input, dim):
    dim_size = symbolic_helper._get_tensor_dim_size(input, dim)
    if dim_size is not None:
        assert dim_size % 2 == 0

     first, second = g.op("Split", input, axis_i=dim, outputs=2)
     return g.op("Mul", first, g.op("Sigmoid", second))
```

torch/onnx/symbolic_opset20.py
```python
 @_onnx_symbolic("aten::glu")
@symbolic_helper.parse_args("v", "i")
@_beartype.beartype
def glu(g: jit_utils.GraphContext, input, dim):
    dim_size = symbolic_helper._get_tensor_dim_size(input, dim)
    if dim_size is not None:
        assert dim_size % 2 == 0

       first, second = g.op("Split", input, axis_i=dim, num_outputs_i=2, outputs=2)
     return g.op("Mul", first, g.op("Sigmoid", second))
```

The code duplication is needed to keep the operator implementation for each opset in their intended `torch/onnx/symbolic_opsetXXX.py` file

Reply:
This code duplication was actually implemented in a previous commit. I found it had too much of code duplication with very little difference between. Please note that not like glu, which is a standalone function, some functions have dependencies which introduce more duplications (_var_mean for example). I consulted with @BowenBao for this approach. I do understand that not having entries in symbolic_opsetxxx.py is confusion. To solve this, shall we move the common code into a helper file and call it from multiple symbolic_opsetxxx.py? I found many sample usages in symbolic_helper.py so I believe it has been the convention for this.

Reply:
regarding:
"Maybe the tests https://github.com/pytorch/pytorch/blob/main/test/onnx/test_utility_funs.py#L276 and https://github.com/pytorch/pytorch/blob/main/test/onnx/test_utility_funs.py#L260 need to be removed?"
Yes I shall have it removed.

Reply:
> This code duplication was actually implemented in a previous commit ([4009b39](https://github.com/pytorch/pytorch/commit/4009b398a02490a4062b46449c76aacd022284ea)). I found it had too much of code duplication with very little difference between. Please note that not like glu, which is a standalone function, some functions have dependencies which introduce more duplications (_var_mean for example). I consulted with @BowenBao for this approach. I do understand that not having entries in symbolic_opsetxxx.py is confusing. To solve this, shall we move the common code into a helper file and call it from multiple symbolic_opsetxxx.py? I found many sample usages in symbolic_helper.py so I believe it has been the convention for this kind of issues.

Indeed "cross-opset" functions were moved as helpers into `symbolic_helper.py` and might be of help for you here

Reply:
I just did with reducemean and reduceprod (https://github.com/pytorch/pytorch/pull/118828/commits/2e5acd4b23580cb0a0e1724957abe7eed27aba43). Will finish the rest.